DOC DB read-only replicas (#587)
silverstripe/developer-docs: en/02_Developer_Guides/08_Performance/07_DB_read_only_replicas.md

Reviewer: Add a link to this page from the environment variables list (https://docs.silverstripe.org/en/5/getting_started/environment_management/#core-environment-variables)

Author: Added

Reviewer: Please give code examples for the new methods on `DB` - there's a lot of new API and no docs to say how to use them?

Author: I've documented `DB::setMustUsePrimary()` and `DB::withPrimary()` as they're both relevant for end-users

I haven't documented any of the other new API such as `DB::hasConfig` and `DB::hasReplicaConfig()` is really only intended for internal use as all the replica stuff happens automatically, all the user needs to do is add the env variables.

---
title: DB read-only replicas
summary: Using DB read-only replicas to improve performance
---

> # DB read-only replicas

Reviewer: ```suggestion
title: Read-only database replicas
summary: Using read-only database replicas to improve performance
---

# Read-only database replicas
```

Author: Updated

Read-only replicas are additional databases that are used to offload read queries from the primary database, which can improve performance by reducing the load on the primary database.

> Read-only replicas are configured by adding environment variables that match the primary environment variable and suffixing `_REPLICA_<replica-number>` to the variable name, where `<replica_number>` is the replica number padding by a zero if it's less than 10, for example `SS_DATABASE_SERVER` becomes `SS_DATABASE_SERVER_REPLICA_01` for the first replica, or `SS_DATABASE_SERVER_REPLICA_12` for the 12th replica. Replias must be numbered sequentially starting from `01`.

Reviewer: This could be understood to mean that you shouldn't have `SS_DATABASE_SERVER` if there's a replica... maybe give a .env example that shows both the primary DB and a single replica, e.g

```
# Primary database
SS_DATABASE_CLASS="MySQLDatabase"
SS_DATABASE_SERVER="my-db-server"
SS_DATABASE_PORT="3306"
SS_DATABASE_USERNAME="my-user"
SS_DATABASE_PASSWORD="my-password"
SS_DATABASE_NAME="db"

# Read-only replica
SS_DATABASE_SERVER_REPLICA_01="my-db-replica"
SS_DATABASE_PORT_REPLICA_01="3306"
SS_DATABASE_USERNAME_REPLICA_01="my-replica-user"
SS_DATABASE_PASSWORD_REPLICA_01="my-replica-password"
```

Author: Added

Replicas cannot define different configuration values for `SS_DATABASE_CLASS`, `SS_DATABASE_NAME`, or `SS_DATABASE_CHOOSE_NAME`. They are restricted to prevent strange issues that could arise from having inconsistent database configurations across replicas.

> If one or more read-only replicas have been configured, then one of the read-only replicas will be selected from the pool of available replicas to handle queries for the rest of the request cycle. However the primary database will be used instead if one of the follow criteria has been met:

Reviewer: ```suggestion
If one or more read-only replicas have been configured, then for each request one of the read-only replicas will be randomly selected from the pool of available replicas to handle queries for the rest of the request cycle. However the primary database will be used instead if one of the follow criteria has been met:
```
Or if it's not random, describe how the selection works.

Author: Updated

- The current query includes any mutable SQL such as `INSERT` or `DELETE`. The primary database will be used for the current query, as well as any future queries, including read queries, for the rest of the current request cycle. Mutable SQL is defined on [`DBConnector::isQueryMutable()`](api:SilverStripe\ORM\Connect\DBConnector::isQueryMutable()).

> - The HTTP request that matched a rule defined in [`Director.must_use_primary_db_rules`](api:SilverStripe\Control\Director->must_use_primary_db_rules). By default the URL paths `Security`, `dev`, and `admin` (if `silverstripe/admin` is installed) are covered by this by default.

Reviewer: ```suggestion
- The HTTP request matches a routing rule defined in [`Director.must_use_primary_db_rules`](api:SilverStripe\Control\Director->must_use_primary_db_rules). By default the URL paths `Security/*`, `dev/*`, and `admin/*` (if `silverstripe/admin` is installed) are covered by this by default.
```

Author: Updated

Reviewer: Still need to add the `/*` so it's clear it's any path starting with those - i.e. `dev/build` is included, not just `dev` on its own.

Author: Updated

> - A user with CMS access is logged in. This is done to ensure that logged in users will correctly see any CMS updates on the website frontend. Users without CMS access will still use the read-only replica.

Reviewer: ```suggestion
- A user with CMS access is logged in. This is done to ensure that logged in users will correctly see any CMS updates on the website frontend. Users without CMS access will still use a read-only replica.
```

Author: Updated

> - The DataObject being queried is configured with [`DataObject.must_use_primary_db`](api:SilverStripe\ORM\DataObject->must_use_primary_db) to `true`, assuming the use of an ORM method that later calls DataQuery::execute(). This includes most commonly used ORM methods such as [`DataObject::get()`](api:SilverStripe\ORM\DataObject::get()), and excludes [`SQLSelect`](api:SilverStripe\ORM\Queries\SQLSelect) methods. By default all core security related DataObjects have `must_use_primary_db` set to `true`.

Reviewer: ```suggestion
- For any query that goes through a call to [`DataQuery::execute()`](api:SilverStripe\ORM\DataQuery::execute()), the `DataObject` subclass being queried is configured with [`DataObject.must_use_primary_db`](api:SilverStripe\ORM\DataObject->must_use_primary_db) set to `true`. This includes most commonly used ORM methods such as [`DataObject::get()`](api:SilverStripe\ORM\DataObject::get()), and excludes [`SQLSelect`](api:SilverStripe\ORM\Queries\SQLSelect) methods. By default all core security related `DataObject` subclasses have `must_use_primary_db` set to `true`.
```

Author: Updated

> - Anything code wrapped in a call to [`DB::withPrimary()`](api:SilverStripe\ORM\DB::withPrimary()).

Reviewer: ```suggestion
- Any code wrapped in a call to [`DB::withPrimary()`](api:SilverStripe\ORM\DB::withPrimary()).
```
This doesn't cause the rest of the request cycle to use the primary DB though, does it? Seems like this should be mentioned outside of this list and include a code block example.

Author: No it doesn't. It does belong in this list though as I've said it's code wrapped in the call, which excludes code not wrapped in it. I'll add a code example

> - CLI is being used instead of HTTP.

Reviewer: ```suggestion
- All queries that result from using the CLI.
```

Author: Updated
